signer/core/signed_data.go

```diff
@@ -0,0 +1,714 @@
+		encType := field["type"]
+		encValue := data[field["name"]]
+		if encType[len(encType)-1:] == "]" {
+			arrayValue, ok := encValue.([]interface{})
+			if !ok {
+				return nil, dataMismatchError(encType, encValue)
+			}
+			typedData.Output.WriteString(strings.Repeat("\u00a0", depth*2))
+			typedData.Output.WriteString(fmt.Sprintf("\"%s\": [ %s\n", field["name"], encType))
+
+			arrayBuffer := bytes.Buffer{}
+			parsedType := strings.Split(encType, "[")[0]
+			for _, item := range arrayValue {
+				if typedData.Types[parsedType] != nil {
+					mapValue, ok := item.(map[string]interface{})
+					if !ok {
+						return nil, dataMismatchError(parsedType, item)
+					}
+					encodedData, err := typedData.EncodeData(parsedType, mapValue, depth+1)
+					if err != nil {
+						return nil, err
+					}
+					arrayBuffer.Write(encodedData)
+				} else {
+					encValue, err := typedData.HandlePrimitiveValue(encType, encValue, depth)
+					if err != nil {
+						return nil, err
+					}
+					bytesValue, err := bytesValueOf(encValue)
+					if err != nil {
+						return nil, err
+					}
+					arrayBuffer.Write(bytesValue)
+				}
+			}
+
+			typedData.Output.Truncate(typedData.Output.Len() - 2)
+			typedData.Output.WriteString(fmt.Sprintf("\n%s],\n", strings.Repeat("\u00a0", depth*2)))
+			encValues = append(encValues, crypto.Keccak256(arrayBuffer.Bytes()))
+		} else if typedData.Types[field["type"]] != nil {
+			mapValue, ok := encValue.(map[string]interface{})
+			if !ok {
+				return nil, dataMismatchError(encType, encValue)
+			}
+			typedData.Output.WriteString(strings.Repeat("\u00a0", depth*2))
+			typedData.Output.WriteString(fmt.Sprintf("\"%s\": { %s\n", field["name"], encType))
+
+			encodedData, err := typedData.EncodeData(field["type"], mapValue, depth+1)
+			if err != nil {
+				return nil, err
+			}
+
+			typedData.Output.Truncate(typedData.Output.Len() - 2)
+			typedData.Output.WriteString(fmt.Sprintf("\n%s},\n", strings.Repeat("\u00a0", depth*2)))
+			encValue = crypto.Keccak256(encodedData)
+			encValues = append(encValues, encValue)
+		} else {
+			primitiveEncValue, err := typedData.HandlePrimitiveValue(encType, encValue, depth)
+			if err != nil {
+				return nil, err
+			}
+			encValues = append(encValues, primitiveEncValue)
+		}
+	}
+
+	buffer := bytes.Buffer{}
+	for _, encValue := range encValues {
+		bytesValue, err := bytesValueOf(encValue)
+		if err != nil {
+			return nil, err
+		}
+		buffer.Write(bytesValue)
+	}
+
+	return buffer.Bytes(), nil // https://github.com/ethereumjs/ethereumjs-abi/blob/master/lib/index.js#L336
+}
+
+// handlePrimitiveValues deals with the primitive values found
+// while searching through the typed data
+func (typedData *TypedData) HandlePrimitiveValue(encType string, encValue interface{}, depth int) (interface{}, error) {
+	var primitiveEncValue interface{}
+
+	typedData.Output.WriteString(strings.Repeat("\u00a0", depth*2))
+	typedData.Output.WriteString(fmt.Sprintf("\"%s\": ", encType))
```

Comment:
I don't like this combination fo parsing and simultaneously building a UI representation. We need to figure out a better way, so that we can serve the UI with something that is easy to display regardless of whether it's displayed in `cli`, rich text or html.

Reply:
Sure thing, I did this only due to performance reasons.